src/raft.rs

```diff
@@ -2483,9 +2599,51 @@ impl<T: Storage> Raft<T> {
             to_send.reject = true;
             to_send.reject_hint = hint_index;
             to_send.log_term = hint_term.unwrap();
+            success = false;
         }
         to_send.set_commit(self.raft_log.committed);
         self.r.send(to_send, &mut self.msgs);
+        success
+    }
+
+    // TODO: revoke pub when there is a better way to test.
+    /// For a given message, append the entries to the log.
+    pub fn handle_append_entries(&mut self, m: &Message) {
+        self.try_append_entries(m);
+    }
+
+    /// For a broadcast, append entries to local log and forward MsgAppend to other dest.
+    pub fn handle_group_broadcast(&mut self, m: &Message) {
+        if self.try_append_entries(m) {
+            // If the agent fails to append entries from the leader,
+            // the agent cannot forward MsgAppend.
+            for forward in m.get_forwards() {
+                self.r
+                    .send_forward(m.from, m.commit, m.commit_term, forward, &mut self.msgs);
+            }
+        } else {
+            for forward in m.get_forwards() {
```

Comment:
What if the agent has conflict log sequence?

Reply:
According to Log Matching Property, the agent must have identical log entries up through the last index if the agent appends log entries from leader successfully.

Reply:
This is `else` branch, which means append fail.

Reply:
If append fails, the agent just sends an empty MsgAppend with no entry. Hence, conflict log sequence in the agent will not affect other followers.
If the follower that receives MsgAppend from the agent also has conflict log, this empty MsgAppend will trigger append reject due to mismatch of previous log, and leader can decrement next index. If the follower does not have conflict log, this empty MsgAppend truncates some uncommitted log entries in the worst case.

Reply:
> If append fails, the agent just sends an empty MsgAppend with no entry.

Why? For example, the leader has (2, 3), (3, 4), (4, 4). The agent has (2, 3), (3, 3), (4, 3), (5, 3). Leader wants to send logs [2, 4) to a follower via this agent. Agent may fail to append the log, but it can send its conflict entries to follower.

Reply:
If the agent has conflict log, it will fail to append. Then it goes to `else` branch. In `else` branch, the agent only sends an empty MsgAppend. 
In this example, if leader wants to send logs [4,5], the previous log in MsgGroupBroadcast is (5, 4) which is a check for the agent, and previous log in Forward is (3, 4) which is the beginning of log entries we want to send via agent. In other words, MsgGroupBroadcast is a MsgAppend with several Forward information. The part of MsgAppend guarantees the log consistency of agent, and Forward information records the range we want to send via agent.
When the agent receives MsgGroupBroadcast, it will first check whether the previous log (5, 4) in MsgGroupBroadcast and (5, 4) mismatches with agent's log. So the agent fail to append the log and goes to `else` branch which only sends empty MsgAppend.

Reply:
If the agent is requesting snapshot, it will always reject the logs. Because the message append is forwarded to follower, so response will be sent back to leader. Leader will retry forwarding again. Hence dead loop.

Reply:
If the agent is requesting snapshot, the progress state is `Snapshot`. And only peers in state `Replicate` will be involved with follower replication. If a peer is in other progress state, messages will be sent in leader replication. https://github.com/LintianShi/tikv/blob/80968da659f04bc172db8fbf497dbe4f034387ce/components/raftstore/src/store/peer.rs#L1953

Reply:
The progress state is not `Snapshot`. When request snapshot explicitly, the state will be `Prob`. How can the library user possibly know the constraint?

Reply:
Maybe we should add more comment on `fn handle_groupbroadcast` We can also write all things in RFC and give the link in comment. Since the whole processing of follower replication is separated in TiKV and raft-rs, it is a bit obscure.

Reply:
The implementation should be more defensive.

Reply:
What does 'defensive' mean? We need double-check in case of the library user possibly break the constraint?

Reply:
Yes. We need to expose usable API and validate the input.